File: LAB4/task_2/ArrayOperations.go
/*
Package main реализует задание по работе с массивами:
1. Генерация массива из n целых чисел (n >= 10) в диапазоне [-100, 100].
2. Вычисление средних арифметических положительных и отрицательных элементов.
   Обмен местами максимального и минимального элементов.
3. Поиск самой длинной убывающей последовательности.
4. Вывод массива N×5, начиная с 1000.
*/

package main

import (
    "fmt"
    "math/rand"
    "time"
)

// GenerateArray генерирует массив случайных целых чисел в диапазоне [-100, 100].
// n: размер массива (n >= 10).
func GenerateArray(n int) []int {
    rand.Seed(time.Now().UnixNano())
    arr := make([]int, n)
    for i := range arr {
        arr[i] = rand.Intn(201) - 100
    }
    return arr
}

// ProcessArray вычисляет средние значения и меняет местами max/min.
// Возвращает средние и изменённый массив.
func ProcessArray(arr []int) (float64, float64, []int) {
    sumPos, countPos := 0, 0
    sumNeg, countNeg := 0, 0

    for _, num := range arr {
        if num > 0 {
            sumPos += num
            countPos++
        } else if num < 0 {
            sumNeg += num
            countNeg++
        }
    }

    avgPos := 0.0
    if countPos > 0 {
        avgPos = float64(sumPos) / float64(countPos)
    }

    avgNeg := 0.0
    if countNeg > 0 {
        avgNeg = float64(sumNeg) / float64(countNeg)
    }

    // Поиск индексов min и max
    minIdx, maxIdx := 0, 0
    for i, num := range arr {
        if num < arr[minIdx] {
            minIdx = i
        }
        if num > arr[maxIdx] {
            maxIdx = i
        }
    }
    arr[minIdx], arr[maxIdx] = arr[maxIdx], arr[minIdx]

    return avgPos, avgNeg, arr
}

// FindDecreasingSequence находит самую длинную убывающую последовательность.
func FindDecreasingSequence(arr []int) []int {
    current := []int{arr[0]}
    longest := []int{}

    for _, num := range arr[1:] {
        if num < current[len(current)-1] {
            current = append(current, num)
        } else {
            if len(current) > len(longest) {
                longest = current
            }
            current = []int{num}
        }
    }

    if len(current) > len(longest) {
        longest = current
    }

    return longest
}

// PrintFormattedArray выводит массив в формате N×5.
func PrintFormattedArray(N int) {
    start := 1000
    for i := 0; i < N; i++ {
        for j := 0; j < 5; j++ {
            fmt.Printf("%d ", start + i*5 + j)
        }
        fmt.Println()
    }
}

func main() {
    // Пункт 1
    arr := GenerateArray(10)
    fmt.Println("1. Исходный массив:", arr)

    // Пункт 2
    avgPos, avgNeg, processedArr := ProcessArray(arr)
    fmt.Printf("\n2. Среднее положительных: %.2f\n", avgPos)
    fmt.Printf("Среднее отрицательных: %.2f\n", avgNeg)
    fmt.Println("Массив после замены:", processedArr)

    // Пункт 3
    seq := FindDecreasingSequence(processedArr)
    fmt.Println("\n3. Убывающая последовательность:", seq)

    // Пункт 4
    fmt.Println("\n4. Форматированный массив (N=4):")
    PrintFormattedArray(4)
}

File: LAB4/task_2/ArrayOperations.cpp
/**
 * @file
 * @brief Решение задания по работе с массивами:
 * 1. Генерация массива из n целых чисел (n >= 10) в диапазоне [-100, 100].
 * 2. Вычисление средних арифметических положительных и отрицательных элементов.
 *    Обмен местами максимального и минимального элементов.
 * 3. Поиск самой длинной убывающей последовательности.
 * 4. Вывод массива N×5, начиная с 1000.
 */

/**
 * Формат следует общепринятым шаблонам (например, Doxygen или Google Style), что делает его:
 * Универсальным — разработчики легко поймут структуру даже в чужом коде.
 * Парсируемым — инструменты для генерации документации (как Doxygen) автоматически распознают теги @param и @return, создавая красивую документацию.
 */

#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <limits>

using namespace std;

/**
 * Генерирует массив случайных целых чисел в диапазоне [-100, 100].
 * @param n Размер массива (n >= 10).
 * @return Сгенерированный массив.
 */
vector<int> generateArray(int n) {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distr(-100, 100);
    vector<int> arr;
    for (int i = 0; i < n; ++i) {
        arr.push_back(distr(gen));
    }
    return arr;
}

/**
 * Находит средние арифметические положительных и отрицательных элементов.
 * Меняет местами максимальный и минимальный элементы.
 * @param arr Исходный массив.
 * @return Кортеж: (среднее положительных, среднее отрицательных, изменённый массив).
 */
tuple<double, double, vector<int>> processArray(vector<int> arr) {
    int sum_pos = 0, count_pos = 0;
    int sum_neg = 0, count_neg = 0;
    
    for (int num : arr) {
        if (num > 0) {
            sum_pos += num;
            count_pos++;
        } else if (num < 0) {
            sum_neg += num;
            count_neg++;
        }
    }
    
    double avg_pos = count_pos > 0 ? (double)sum_pos / count_pos : 0.0;
    double avg_neg = count_neg > 0 ? (double)sum_neg / count_neg : 0.0;
    
    auto [min_it, max_it] = minmax_element(arr.begin(), arr.end());
    iter_swap(min_it, max_it);
    
    return make_tuple(avg_pos, avg_neg, arr);
}

/**
 * Находит самую длинную убывающую последовательность в массиве.
 * @param arr Исходный массив.
 * @return Вектор, содержащий последовательность.
 */
vector<int> findDecreasingSequence(const vector<int>& arr) {
    vector<int> current = {arr[0]};
    vector<int> longest;
    
    for (size_t i = 1; i < arr.size(); ++i) {
        if (arr[i] < current.back()) {
            current.push_back(arr[i]);
        } else {
            if (current.size() > longest.size()) {
                longest = current;
            }
            current = {arr[i]};
        }
    }
    
    if (current.size() > longest.size()) {
        longest = current;
    }
    
    return longest;
}

/**
 * Выводит массив в формате N×5, начиная с 1000.
 * @param N Количество строк.
 */
void printFormattedArray(int N) {
    int start = 1000;
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < 5; ++j) {
            cout << start + i * 5 + j << " ";
        }
        cout << endl;
    }
}

int main() {
    // Пункт 1
    auto arr = generateArray(10);
    cout << "1. Исходный массив: ";
    for (int num : arr) cout << num << " ";
    cout << endl;

    // Пункт 2
    auto [avg_pos, avg_neg, processed_arr] = processArray(arr);
    cout << "\n2. Среднее положительных: " << avg_pos << endl;
    cout << "Среднее отрицательных: " << avg_neg << endl;
    cout << "Массив после замены: ";
    for (int num : processed_arr) cout << num << " ";
    cout << endl;

    // Пункт 3
    auto seq = findDecreasingSequence(processed_arr);
    cout << "\n3. Убывающая последовательность: ";
    for (int num : seq) cout << num << " ";
    cout << endl;

    // Пункт 4
    cout << "\n4. Форматированный массив (N=4):" << endl;
    printFormattedArray(4);

    return 0;
}
